composer/core/state.py

```diff
@@ -501,6 +525,19 @@ def algorithms(self):
     def algorithms(self, algorithms: Sequence[Algorithm]):
         self._algorithms[:] = algorithms
 
+    def _dataset_of(self, obj: Optional[Union[Iterable, Evaluator, DataSpec, DataLoader]]) -> Optional[Dataset]:
```

Comment:
can you add a comment on how this fn works? I was also confused (like Karan) on how this worked so it'd be helpful to explain the iterative unpacking